azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -0,0 +1,175 @@
+from azurelinuxagent.common import conf
+from azurelinuxagent.common.osutil import get_osutil
+import azurelinuxagent.ga.policy.regorus as regorus
+from azurelinuxagent.ga.policy.regorus import PolicyError
+
+# Define support matrix for Regorus and policy engine feature.
+# Dict in the format: { distro:min_supported_version }
+POLICY_SUPPORTED_DISTROS_MIN_VERSIONS = {
+    'ubuntu': DistroVersion('16.04'),
+    'mariner': DistroVersion('2'),
+    'azurelinux': DistroVersion('3')
+}
+# TODO: add 'arm64', 'aarch64' here once support is enabled for ARM64
+POLICY_SUPPORTED_ARCHITECTURE = ['x86_64']
+
+
+class PolicyEngine(object):
+    """
+    Implements policy engine API. Class will always be initialized, but if the Regorus import fails,
+    all methods will be no-ops.
+
+    If any errors are thrown in regorus.py, they will be caught and handled here. add_policy, add_data,
+    and set_input will be no-ops, eval_query will return an empty dict
+    """
+    def __init__(self, rule_file, policy_file):
+        """
+        rule_file: Path to a Rego file that specifies rules for policy behavior.
+
+        policy_file: Path to a JSON file that specifies parameters for policy behavior - for example,
+        whether allowlist or extension signing should be enforced.
+        The expected file format is:
+        {
+            "azureGuestAgentPolicy": {
+                "policyVersion": "0.1.0",
+                "signingRules": {
+                    "extensionSigned": <true, false>
+                },
+                "allowListOnly": <true, false>
+            },
+            "azureGuestExtensionsPolicy": {
+                "allowed_ext_1": {
+                    "signingRules": {
+                        "extensionSigned": <true, false>
+                    }
+                }
+        }
+        """
+        self._policy_engine_enabled = False
+        self._engine = None
+        if not self.is_policy_enforcement_enabled():
+            self.log_policy(msg="Policy enforcement is not enabled.")
+            return  # policy_engine_enabled is not set to True
+
+        # If unsupported, this call will raise an error
+        self.check_policy_enforcement_supported()
+
+        try:
+            self._engine = regorus.Engine(policy_file=policy_file, rule_file=rule_file)
+            self._policy_engine_enabled = True
+        except Exception as ex:
+            msg = "Failed to initialize Regorus policy engine: '{0}'".format(ex)
+            self.log_policy(msg=msg)
+            raise PolicyError(msg)
+
+    @classmethod
+    def log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_event=True):
+        """
+        Log information to console and telemetry.
+        """
+        if is_success:
+            logger.info(msg)
+        else:
+            logger.error(msg)
+        if send_event:
+            add_event(op=op, message=msg, is_success=is_success)
+
+    @staticmethod
+    def is_policy_enforcement_enabled():
```

Comment:
should this method be private, or are we expecting callers to check this before invoking other methods in PolicyENgine?

Reply:
similar question for check_policy_enforcement_supported()

Reply:
I see, looks like policy_engine_enabled() is a left-over from a previous iteration of the code and that the other 2 methods are invoked from init() and should be private

Reply:
policy_engine_enabled( ) will be used in the future by ExtensionPolicyEngine, in the function is_extension_download_allowed( ). 

The other two methods can be made private, since ExtensionPolicyEngine can just check the policy_engine_enabled propety.

Reply:
thanks